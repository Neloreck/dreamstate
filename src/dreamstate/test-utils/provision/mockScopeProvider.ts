import { Provider } from "react";

import { IScopeContext, ScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

/**
 * Mock scope with initial clean state.
 * Can be used to mock whole scope context.
 *
 * Should be used in pair with 'mockScope' method so 'value' property of ScopeContext.Provider will be correct.
 *
 * @returns {Provider} ScopeContext.Provider component for react rendering.
 */
export function mockScopeProvider(): Provider<IScopeContext> {
  return ScopeContext.Provider;
}
