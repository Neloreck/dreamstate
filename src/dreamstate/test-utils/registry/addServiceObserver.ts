import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { addServiceObserverToRegistry } from "@/dreamstate/core/registry/addServiceObserverToRegistry";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import { TAnyContextServiceConstructor, TUpdateObserver } from "@/dreamstate/types";

/**
 * Add context manager observer and trigger all related events (onProvisionStarted for first observer).
 */
export function addServiceObserver(
  Service: TAnyContextServiceConstructor,
  observer: TUpdateObserver
) {
  if (!(Service.prototype instanceof ContextService)) {
    throw new TypeError("Supplied class is not extending ContextService.");
  }

  if (!CONTEXT_SERVICES_REGISTRY.get(Service)) {
    throw new Error("Supplied class was not registered.");
  }

  addServiceObserverToRegistry(Service, observer);
}
