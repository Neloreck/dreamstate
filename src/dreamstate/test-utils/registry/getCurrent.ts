import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TAnyContextManagerConstructor>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
