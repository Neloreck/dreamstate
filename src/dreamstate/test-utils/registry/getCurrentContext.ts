import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current supplied context.
 */
export function getCurrentContext<T extends TAnyContextManagerConstructor>(
  Manager: T
): T["prototype"]["context"] | null {
  // todo; return CONTEXT_STATES_REGISTRY.get(Manager) || null;
  return null as any;
}
