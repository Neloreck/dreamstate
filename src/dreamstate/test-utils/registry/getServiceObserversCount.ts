import { CONTEXT_OBSERVERS_REGISTRY } from "@/dreamstate/core/internals";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { TAnyObject, TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current service observers count.
 * Observers - all points that create react context provision.
 */
export function getServiceObserversCount<S extends TAnyObject, T extends TAnyContextManagerConstructor>(
  Service: T
): number | null {
  if (!(Service.prototype instanceof ContextManager)) {
    throw new TypeError("Supplied parameter should be class extending ContextManager.");
  }

  if (!CONTEXT_OBSERVERS_REGISTRY.has(Service)) {
    throw new Error("Supplied class was not registered.");
  }

  return CONTEXT_OBSERVERS_REGISTRY.get(Service)!.size;
}
