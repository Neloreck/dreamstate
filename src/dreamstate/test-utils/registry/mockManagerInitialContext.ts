import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Mock manager initial context on service registration.
 * Packs manager and context in map in a type-safe way instead of manual way.
 *
 * @param ManagerClass - manager class to provide initial context to.
 * @param context - initial context partial for applying on provision.
 */
export function mockManagerInitialContext<D extends TAnyContextManagerConstructor>(
  ManagerClass: D,
  context: Partial<D["prototype"]["context"]>
): [D, Partial<D["prototype"]["context"]>] {
  return [ ManagerClass, context ];
}
