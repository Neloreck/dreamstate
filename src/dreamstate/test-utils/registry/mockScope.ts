import { initializeScopeContext } from "@/dreamstate/core/scoping/initializeScopeContext";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

/**
 * Mock scope with clean initial state.
 * Can be used to mock whole scope context and run some advanced tests.
 *
 * @returns {IScopeContext} mocked scope context.
 */
export function mockScope(): IScopeContext {
  return initializeScopeContext();
}
