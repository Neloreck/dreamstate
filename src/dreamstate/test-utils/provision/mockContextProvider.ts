import { createElement, FunctionComponent, Provider, ReactNode } from "react";

import { createProvider, ICreateProviderProps } from "@/dreamstate/core/provision/createProvider";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { mockScopeProvider } from "@/dreamstate/test-utils/provision/mockScopeProvider";
import { mockScope } from "@/dreamstate/test-utils/registry/mockScope";
import { IContextManagerConstructor, TAnyObject } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Mock context provider that does not fire provision event and just constructs instances.
 * Can be used to mock tree provision where required.
 *
 * @param {Array.<IContextManagerConstructor>>} sources - array of source classes references
 *   that should be provided in react tree when returned component renders.
 * @param {ICreateProviderProps} config - store provider configuration.
 * @param {boolean} config.isCombined - boolean flag that switches observing between one big react node
 *   vs small scoped nodes.
 * @param {IScopeContext} scope - scope where providers should be injected, mocked by default.
 * @returns {FunctionComponent} mocked react provider component for source classes in mocked scope.
 */
export function mockContextProvider<T extends TAnyObject = TAnyObject>(
  sources: Array<IContextManagerConstructor>,
  config: ICreateProviderProps = {},
  scope: IScopeContext = mockScope()
): FunctionComponent<IProviderProps<T>> {
  const scopeProviderProps: { value: IScopeContext } = { value: scope };
  const ContextProvider: FunctionComponent<IProviderProps<T>> = createProvider(sources, config);
  const ScopeProvider: Provider<IScopeContext> = mockScopeProvider();

  /**
   * Create provider component that can be used by react to automatically inject specific scope.
   */
  function MockedProvider(props: T): ReactNode {
    return createElement(ScopeProvider, scopeProviderProps, createElement(ContextProvider, props));
  }

  /**
   * Indicate that following provider is mocked.
   */
  MockedProvider.displayName = "MockedProvider";

  return MockedProvider as FunctionComponent<IProviderProps<T>>;
}
