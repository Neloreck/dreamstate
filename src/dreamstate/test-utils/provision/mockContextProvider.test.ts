import { render } from "enzyme";
import { createElement } from "react";

import { useManager } from "@/dreamstate";
import { mockContextProvider } from "@/dreamstate/test-utils/provision/mockContextProvider";
import { mockManagerInitialContext } from "@/dreamstate/test-utils/registry/mockManagerInitialContext";
import { mockScope } from "@/dreamstate/test-utils/registry/mockScope";
import { TestManager } from "@/fixtures";

describe("mockContextProvider util", () => {
  it("should correctly provide mocked state for nested components", () => {
    const Provider = mockContextProvider(
      [TestManager],
      {},
      mockScope({
        applyInitialContexts: [mockManagerInitialContext(TestManager, { first: "this test", second: 1000 })],
      })
    );

    const SomeComponent = ({ testContext: { first, second, third } = useManager(TestManager) }) => {
      expect(first).toBe("this test");
      expect(second).toBe(1000);
      expect(third).toBe(false);

      return createElement("div", {}, [
        createElement("span", { key: 1 }, first),
        createElement("span", { key: 2 }, second),
        createElement("span", { key: 3 }, JSON.stringify(third)),
      ]);
    };

    const tree = render(createElement(Provider, {}, createElement(SomeComponent)));

    expect(tree).toMatchSnapshot();
  });
});
