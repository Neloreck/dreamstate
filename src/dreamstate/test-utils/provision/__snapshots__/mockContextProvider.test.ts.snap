// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mockContextProvider util should correctly provide mocked state for nested components 1`] = `
<div>
  <span>
    this test
  </span>
  <span>
    1000
  </span>
  <span>
    false
  </span>
</div>
`;
